Add features into the multipart upload.

melgenek opened the pull request:
There are multiple changes in this pr:
1) The chunk size is configurable.
The chunk size can be other than 5 mb.
https://docs.aws.amazon.com/AmazonS3/latest/dev/UploadingObjects.html
```
...  You can use a multipart upload for objects from 5 MB to 5 TB in size. ...
``` 
There can be at most 10000 parts, so the current implementation has 5 mb * 10000 = 50 GB. With the configurable chunk size, the maximum file size can be bigger.

2) The chunks are uploaded in parallel.
The default value is set to 1 to keep the current behaviour with the default configuration. Parallel requests can improve the speed of the file upload.

3) It's possible to pass the canned acl during the multipart upload.
In the cross-account scenario, the owner of the bucket doesn't have the access to the objects in the bucket by default. The change allows supporting the scenario that is described in https://docs.aws.amazon.com/AmazonS3/latest/dev/amazon-s3-policy-keys.html#grant-putobject-conditionally-1.
Unfortunately, the test to check that the acl is applied is not provided because minio does not support the object-level acl policies 
https://docs.minio.io/docs/minio-server-limits-per-tenant.html
```
List of Amazon S3 Object API's not supported on MinIO
ObjectACL (Use bucket policies instead)
```
The default value is "private". It is assigned if no values are provided https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html#canned-acl

|Canned ACL | Applies to              | Permissions added to ACL |
| ------------- |-------------| -----|
|private           | Bucket and object | Owner gets FULL_CONTROL. No one else has access rights (default).|

4)  Don't fail when the stream is empty.
Currently, the upload fails when the stream has no bytes.

The `completeMultipartUpload` produces a not really clear exception
```
             software.amazon.awssdk.services.s3.model.S3Exception: The XML you provided was not well-formed or did not validate against our published schema. (Service: S3, Status Code: 400, Request ID: 163337F34D7A7576, Extended Request ID: null)
            	at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handleErrorResponse(AwsXmlPredicatedResponseHandler.java:156)
            	at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handleResponse(AwsXmlPredicatedResponseHandler.java:106)
            	at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handle(AwsXmlPredicatedResponseHandler.java:84)
            	at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handle(AwsXmlPredicatedResponseHandler.java:42)
            	at software.amazon.awssdk.core.internal.handler.BaseClientHandler.lambda$successTransformationResponseHandler$4(BaseClientHandler.java:214)
            	at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler.lambda$prepare$0(AsyncResponseHandler.java:88)
            	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072)
            	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
            	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
            	at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler$BaosSubscriber.onComplete(AsyncResponseHandler.java:129)
            	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.runAndLogError(ResponseHandler.java:179)
            	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.access$500(ResponseHandler.java:69)
            	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler$PublisherAdapter$1.onComplete(ResponseHandler.java:295)
            	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerPublisher.publishMessage(HandlerPublisher.java:402)
            	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerPublisher.flushBuffer(HandlerPublisher.java:338)
            	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerPublisher.receivedDemand(HandlerPublisher.java:291)
            	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerPublisher.access$200(HandlerPublisher.java:61)
            	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerPublisher$ChannelSubscription$1.run(HandlerPublisher.java:495)
            	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
            	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
            	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:497)
            	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
            	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
            	at java.base/java.lang.Thread.run(Thread.java:834)
```

This exception happens because the `parts` list is empty.
This pr aborts the operation instead of throwing the exception. This results in a successfully completed upload without any files ending up in the s3.
An alternative solution is to put an empty file in s3. Alpakka uses this approach.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/zio/zio-s3?pullRequest=117) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/zio/zio-s3?pullRequest=117) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/zio/zio-s3?pullRequest=117) it.</sub>

regis-leray commented:
Hello, thanks for you contribution

1. Yeah chunk size should be configurable you are totally right 

i would more do something like this to remove too many parameters 

```scala
override def multipartUpload[R <: zio.Has[_]: Tag](
        bucketName: String,
        key: String,
        contentType: String,
        content: ZStream[R, Throwable, Byte],
        options: MultipartUploadOptions = MultipartUploadOptions.default
      ): ZIO[R, S3Exception, Unit] 



sealed trait ChunkSize {
  def size: Int
}

object ChunkSize {
  //part size limit is 5Mb, required by amazon api
  val Min = 5 * 1024 * 1024

  val default = new ChunkSize {
    val size: Int = Min
  }

  def from(_size: Int): Either[String, ChunkSize] =
    _size match {
      case s if s >= Min =>
        Right(value = new ChunkSize {
          val size: Int = _size
        })
      case invalid       => Left(s"Invalid chunk size $invalid, minimum value is $Min ")
    }
}

case class MultipartUploadOptions(metadata: Map[String, String], chunkSize: ChunkSize, acl: ObjectCannedACL)

object MultipartUploadOptions {
  val default: MultipartUploadOptions = MultipartUploadOptions(Map.empty, ChunkSize.default, ObjectCannedACL.PRIVATE)
}
```

2. Concerning parrallel of upload, did you do some benchmarking on real amazon s3. I did it my self , and they were no improvement even worse bad perf
https://github.com/zio/zio-s3/pull/10

3. see proposal 1

4. your fix looks good,

regis-leray commented:
need to investigate why circleci dont work anymore

regis-leray reviewed (changes requested):
please see my comments

regis-leray commented:
if you update and push new commits to your PR, it should retrigger circleci

melgenek commented:
@regis-leray hi, thank you for the review!

1) I removed the chunk size validation. The aws client provides a proper S3Exception 
```
  software.amazon.awssdk.services.s3.model.S3Exception: Your proposed upload is smaller than the minimum allowed size (Service: S3, Status Code: 400, Request ID: 66A07B7C2866ECE5, Extended Request ID: NRDAc2xG7ClhhDXAl9uJCQpio95f8arYqVw7kNY4j+6q7oXwr5DPBoJe9zCtdVsOILYOCAOQI14=)
	at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handleErrorResponse(AwsXmlPredicatedResponseHandler.java:156)
....
```


2) I added a test file for uploading in parallel. The results on my machine are
```
Parallelism 4 58265
Parallelism 2 68993
Parallelism 1 144218

Parallelism 4 45566
Parallelism 2 67623
Parallelism 1 145243

Parallelism 4 61833
Parallelism 2 75532
Parallelism 1 232915
```
It's visible that the parallelism of 2 gives about 2 times speed increase compared to 1. Not sure why 4 is not too much better than 2, but I can assume that my laptop doesn't give more resources for the upload. In general, you can see that there is a clear benefit in the speed of upload. Tweaking the chunk sizes can give even better upload times.

3) I could add the MultipartUploadOptions. However, the content type should also be added there, because it is in fact optional. Both content type and metadata have already been in the signature so this change would not be invisible for those who upgrade. If it's fine, then I'll introduce the options.

regis-leray commented:
@melgenek hello

1. yeah this exception totally fine

2. did you test with a real amazon s3 bucket or Minio ? my file what 60 Mb
could also try with `mapPar(20)`

maybe they change something in ZIO, i didnt test it since zio 1.0 :)


3. yes you could move ContentType as Option[String] in MultipartUploadOptions

melgenek commented:
The parallel upload was done against the real s3. I guess my computer won't show any benefit on 20 parallel uploads. I tried at most 8 parallel requests. It also gave a small insignificant improvement in speed.

I will make the changed to the config and empty stream handling a bit later.

regis-leray commented:
@melgenek  thanks a lot

melgenek commented:
This pr is getting bigger than I hoped it to be :slightly_smiling_face: 

The empty stream behaviour is updated to create an empty file.

I started pulling out the upload options and turned out that the test service is passing these options into the "put request". So what I did is a class UploadOptions that contains the shared options for both the multipart upload and the put request. They are in essence the same, only the actual execution differs. So the method signatures kept the specifics: (1) the "put" request keeps the content length and (2) the "multipart" upload has the parallelism and the part size.

The "chunk" size is renamed to be the "part" size because the upload is actually "multi**part**".

While writing tests I actually wanted to make sure that the content type and the metadata is being set properly. For this I needed metadata. So I added a simple implementation of the getObjectMetadata with the very basic parameters and used it in the tests.

regis-leray commented on src/main/scala/zio/s3/Live.scala:
```diff
@@ -105,78 +108,78 @@ final class Live(unsafeClient: S3AsyncClient) extends S3.Service {
+                         )
+                       ).map(_.uploadId())
+
+      parts         <- content
```
that should handle the empty case, need to test with amazon api if you can send send request  with same part number 

(ZStream.succeed(Chunk.empty -> 0L) ++ content
                  //part size limit is 5Mb, required by amazon api
                    .chunkN(5 * 1024 * 1024)
                    .mapChunks(Chunk.single)
                    .zipWithIndex)

melgenek replied:
the problem with this approach is the following: if the stream is actually not empty, it won't work.
For example, we have a real chunk of size 1MB. And we prepend 0 bytes chunk. So now we have 2 chunks: [0, 1Mb].
The last chunk can be smaller than 5MB. But while we have 2 chunks, so the call will send an exception, because the first chunk is also smaller than 5 MB.

Akka streams has a nice operator for this https://doc.akka.io/docs/akka/current/stream/operators/Source-or-Flow/orElse.html. But I haven't found a similar operator in zio.

regis-leray replied:
- if stream is empty we are sending an empty byte (same as your solution)  => s3 sending OK
- if stream is not empty inferior of 5Mb, we are sending 0bytes, and after 1Mb => s3 sending FAIL when 1Mb is received
- if stream is 6Mb, we are sending 0bytes, and after 6Mb => s3 sending OK

are we agree ?

melgenek replied:
> if stream is not empty inferior of 5Mb, we are sending 0bytes with PartNumber 1, and after 1Mb with PartNumber 1 => s3 sending FAIL when 1Mb is received
yes

> if stream is 6Mb, we are sending 0bytes with PartNumber 1, and after 6Mb with PartNumber 1 => s3 sending OK
the problem here is that 6MB is actually split into 2 chunks: 1MB and 5MB

regis-leray replied:
sorry i wanted to say 

if stream is 6Mb, we are sending 0bytes with PartNumber 1, 5Mb with PartNumber 1 and 1Mb with partNumber 2 => s3 sending OK

melgenek replied:
here is an implementation without an explicit call https://github.com/zio/zio-s3/pull/117/commits/d6b2f1b0bb06dfc48d245b6409ab4654c6479d82

regis-leray commented on src/main/scala/zio/s3/Live.scala:
```diff
@@ -105,78 +108,78 @@ final class Live(unsafeClient: S3AsyncClient) extends S3.Service {
+    options: UploadOptions = UploadOptions.default,
+    partSize: Int = UploadOptions.MinMultipartPartSize,
+    parallelism: Int = 1
   ): ZIO[R, S3Exception, Unit] =
```
i would be more in favor to keep this definition

```scala
 def multipartUpload[R <: zio.Has[_]: Tag](
    bucketName: String,
    key: String,
    content: ZStream[R, Throwable, Byte],
    options: UploadOptions = MultiPartUploadOptions.default)(parallelism: Int = 1) 
```

melgenek replied:
why is parallelism different compared to the other options? And where should the "partSize" be placed in this case?

regis-leray replied:
i prefer to separate to make it more obvious this is a special parameter. parallelism can have a big impact on performance/behavior of your application same as

```scala
mapMPar(n: Int)(f: O => ZIO[R1, E1, O2])
```

melgenek replied:
I guess that's because the last arguments can become a code block
```
mapMPar(4) { value =>
   ....
}
```

regis-leray replied:
yep mapMPar have a better syntax :)

```scala
val content: ZStream[Any, Nothing, String] = ???

// more obvious 10 is a special value
multipartUpload("bucket1", "key1", s)(10)

//les obvious
multipartUpload("bucket1", "key1", s, 10)
```

melgenek replied:
while adding the multipart specific options it seemed to me that putting parallelism there looks nice. So the parallelism is now considered a specific option as well. See https://github.com/zio/zio-s3/pull/117/commits/7f6632881537df013d77038a4e2f486a2a1dbb7a

regis-leray commented on src/main/scala/zio/s3/Live.scala:
```diff
@@ -105,78 +108,78 @@ final class Live(unsafeClient: S3AsyncClient) extends S3.Service {
+                       )
     } yield ()
 
+  private def uploadPart(bucketName: String, key: String, partNumber: Int, uploadId: String, chunk: Chunk[Byte]) =
```
from my solution you dont need to exrtact this method

melgenek replied:
https://github.com/zio/zio-s3/pull/117/commits/d6b2f1b0bb06dfc48d245b6409ab4654c6479d82

regis-leray commented on src/main/scala/zio/s3/Test.scala:
```diff
@@ -71,6 +70,12 @@ object Test {
           .mapError(S3ExceptionUtils.fromThrowable)
           .provide(blocking)
 
+      override def getObjectMetadata(bucketName: String, key: String): IO[S3Exception, ObjectMetadata] =
```
👍

regis-leray commented on src/main/scala/zio/s3/UploadOptions.scala:
```diff
@@ -0,0 +1,42 @@
+  cannedAcl: ObjectCannedACL = ObjectCannedACL.PRIVATE,
+  contentType: Option[String] = None
+)
+
```
```scala
case class UploadOptions(
  metadata: Map[String, String],
  cannedAcl: ObjectCannedACL
  contentType: Option[String]
)

object UploadOptions {
  val default: UploadOptions = MultipartUploadOptions(Map.empty, ChunkSize.default, ObjectCannedACL.PRIVATE)
}

sealed trait ChunkSize {
  def size: Int
}

object ChunkSize {
  //part size limit is 5Mb, required by amazon api
  val Min = 5 * 1024 * 1024

  val default = new ChunkSize {
    val size: Int = Min
  }

  def from(_size: Int): Either[String, ChunkSize] =
    _size match {
      case s if s >= Min =>
        Right(value = new ChunkSize {
          val size: Int = _size
        })
      case invalid       => Left(s"Invalid chunk size $invalid, minimum value is $Min ")
    }
}

case class MultipartUploadOptions(upload; UploadOptions, chunkSize: ChunkSize)

object MultipartUploadOptions {
  val default: MultipartUploadOptions = MultipartUploadOptions(UploadOptions.default, ChunkSize.default)
}

```

melgenek replied:
looks good to me.
Can we keep the chunk size as Int? 
It would be easier to use because no additional steps except for the multipart upload call is needed. And as I mentioned above there is a dedicated S3Exception produced when the chunk size is outside the valid range.

regis-leray replied:
I would prefer to handle as a sealed trait, like this user would have to understand why this Either :)

melgenek replied:
then the "parallelism" has to also be a type because it can be -1, for example. Obviously there is a check inside zio that doesn't allow negative parallelism, but zio's `mapMPar` still accepts a plain Int

regis-leray replied:
true !
you can validate this number inside the method :) if you want.
I let you decide. my choice would be do nothing ;p

melgenek replied:
there is a dedicated config now https://github.com/zio/zio-s3/pull/117/commits/7f6632881537df013d77038a4e2f486a2a1dbb7a

regis-leray commented on src/main/scala/zio/s3/MultipartUploadOptions.scala:
```diff
@@ -0,0 +1,35 @@
+
+  final val default = MultipartUploadOptions()
+
+  def multipartUploadBuilder(options: MultipartUploadOptions): CreateMultipartUploadRequest.Builder = {
```
You dont need this builder, case class provide already this

melgenek replied:
done in this commit https://github.com/zio/zio-s3/pull/117/commits/bda1d842c0e9c8f75efb4e3e1084fc3270b79a96

regis-leray commented on src/main/scala/zio/s3/MultipartUploadOptions.scala:
```diff
@@ -0,0 +1,35 @@
+ * @param parallelism the number of parallel requests to upload chunks, default to 1
+ */
+case class MultipartUploadOptions(
+  uploadOptions: UploadOptions = UploadOptions.default,
```
it is usually not a good practice to set default value in case class, better to do it companion object

regis-leray commented on src/main/scala/zio/s3/UploadOptions.scala:
```diff
@@ -0,0 +1,31 @@
+ * @param cannedAcl a canned acl, defaults to "private"
+ * @param contentType the content type of the object (application/json, application/zip, text/plain, ...)
+ */
+case class UploadOptions(
```
can you move this into settings.scala

melgenek replied:
https://github.com/zio/zio-s3/pull/117/commits/bda1d842c0e9c8f75efb4e3e1084fc3270b79a96

regis-leray commented on src/main/scala/zio/s3/UploadOptions.scala:
```diff
@@ -0,0 +1,31 @@
+case class UploadOptions(
+  metadata: Map[String, String] = Map.empty,
+  cannedAcl: ObjectCannedACL = ObjectCannedACL.PRIVATE,
+  contentType: Option[String] = None
```
better to declare default value in companion object

regis-leray commented on src/main/scala/zio/s3/UploadOptions.scala:
```diff
@@ -0,0 +1,31 @@
+object UploadOptions {
+  final val default = UploadOptions()
+
+  def putObjectBuilder(options: UploadOptions): PutObjectRequest.Builder = {
```
you dont need builder, you can move it back were it is only use

melgenek replied:
done in this commit https://github.com/zio/zio-s3/pull/117/commits/bda1d842c0e9c8f75efb4e3e1084fc3270b79a96

regis-leray commented on src/main/scala/zio/s3/Live.scala:
```diff
@@ -105,76 +108,71 @@ final class Live(unsafeClient: S3AsyncClient) extends S3.Service {
+                                 )
+                               ).map(_.uploadId())
+
+      chunkedStream          = content.chunkN(options.partSize).mapChunks(Chunk.single).mapError(S3ExceptionUtils.fromThrowable)
```
i  better like this approach, but i was wondering if my solution was working with amazon s3, if not your solution is totaly valid, was not aware of peel. good to know :)

regis-leray replied:
oki i tested my solution locally, and amazon s3 don't like it.

melgenek replied:
https://github.com/zio/zio-s3/pull/117/commits/a1067926d148b8d2bf79a520fa5771945a73510c

regis-leray commented on src/main/scala/zio/s3/Live.scala:
```diff
@@ -237,4 +235,7 @@ object Live {
     }
   }
+
+  final val SingleEmptyChunkStream = ZStream(Chunk.empty)
```
dont need to extract

melgenek replied:
https://github.com/zio/zio-s3/pull/117/commits/a1067926d148b8d2bf79a520fa5771945a73510c#diff-4371b13eba4391eed947c00dddac1f12L239

regis-leray commented on src/main/scala/zio/s3/MultipartUploadOptions.scala:
```diff
@@ -0,0 +1,35 @@
+ */
+case class MultipartUploadOptions(
+  uploadOptions: UploadOptions = UploadOptions.default,
+  partSize: Int = MinMultipartPartSize,
```
please put back the sealed trait

melgenek replied:
Sorry in advance for the long read :smile: Here I am trying to explain why, I think, having no additional types other than Int would be my choice.

I thought that this comment was about not adding the sealed trait https://github.com/zio/zio-s3/pull/117#discussion_r487325347

In the example above that, you showed, there is a ChunkSize type
```
sealed trait ChunkSize {
  def size: Int
}
object ChunkSize {
  def from(size: Int): Either[String, ChunkSize] = ???
}
```
Most likely it is going to be used the way, that the chunk is validated an is immediately passed into the upload
```
  val result: ZIO[S3, io.Serializable, Unit] = for {
    partSize <- ZIO.fromEither(ChunkSize.from(100))
    _ <- multipartUpload("bucket", "key", ZStream.empty, MultipartUploadOptions(partSize = partSize))
  } yield ()
```
Here is the first problem: the error type is Serializable, because the multipartUpload's error is S3Exception, but the ChunkSize's error is a String.

Let's say we use the S3Exception in the "from" method (I also rename the ChunkSize to the PartSize, because that's what we use in the other names):
```
sealed trait PartSize {
  def size: Int
}
object PartSize {
  def from(size: Int): Either[S3Exception, PartSize] = ???
}
```
Now the error type is S3Exception everywhere.

But there is the problem that I mentioned earlier: the API is not consistent, because the "parallelism" is also a potentially invalid value but we pass it as a plain Int. This is what happens if we create Parallelism class and use it
```
  val result = for {
    partSize <- ZIO.fromEither(PartSize.from(100))
    parallelism <- ZIO.fromEither(Parallelism.from(4))
    _ <- multipartUpload("bucket", "key", ZStream.empty, MultipartUploadOptions(partSize = partSize))
  } yield ()
```
Everything compiles, the error is fine, we are sure that the values of parallelism and the partSize are valid. But the user is now responsible for validating all those values. If there is another option added that requires validation, the user would have to add another line with the ".from" statement.

If you really want to have the MultipartUploadOptions that have the fields as custom types rather than Ints, I would create the build for the options and make the case classe's constructor private like this
```
final case class PartSize(value: Int)    extends AnyVal
final case class Parallelism(value: Int) extends AnyVal

case class MultipartUploadOptions private (
  uploadOptions: UploadOptions,
  partSize: PartSize,
  parallelism: Parallelism
)

object MultipartUploadOptions {
  final val MinPartSize: Int = 5 * 1024 * 1024 // 5 MB

  def make(
    uploadOptions: UploadOptions = UploadOptions(),
    partSize: Int = MinPartSize,
    parallelism: Int = 1
  ): Either[S3Exception, MultipartUploadOptions] = ???
}
```

This way creating the options is always a single additional line, even if new options are added
```
  val result = for {
    options <- ZIO.fromEither(MultipartUploadOptions.make(partSize = 1000, parallelism = 4))
    _ <- multipartUpload("bucket", "key", ZStream.empty, options)
  } yield ()
```


Although having types instead of Ints sounds reasonable to me, as a user of the library I would prefer the library doing as much as possible for me. So it would be great if this additional line to create the options can be removed. After all, I will receive the S3Exception anyway. That's why I would prefer having Ints in the MultipartUploadOptions and don't the validation directly in "multipartUpload" method.

regis-leray commented on src/main/scala/zio/s3/MultipartUploadOptions.scala:
```diff
@@ -0,0 +1,35 @@
+ * @param partSize the size of the part in bytes, the minimum is 5 MB
+ * @param parallelism the number of parallel requests to upload chunks, default to 1
+ */
+case class MultipartUploadOptions(
```
can you move this in settings.scala

melgenek replied:
https://github.com/zio/zio-s3/pull/117/commits/bda1d842c0e9c8f75efb4e3e1084fc3270b79a96#diff-362a387518f51a267b646add5682a0d6R33

regis-leray commented on src/test/scala/zio/s3/S3Test.scala:
```diff
@@ -22,20 +22,24 @@ object S3LiveSpec extends DefaultRunnableSpec {
-  override def spec = S3Suite.spec("S3LiveSpec", root).provideCustomLayerShared(s3)
+  override def spec =
+    suite("S3LiveSpec")(
+//      S3Suite.spec("Common spec", root),
```
remove comment

melgenek replied:
https://github.com/zio/zio-s3/pull/117/commits/a1067926d148b8d2bf79a520fa5771945a73510c#diff-8e768d82cb4c34d613df68010821e7daR27

regis-leray commented on src/main/scala/zio/s3/UploadOptions.scala:
```diff
@@ -0,0 +1,31 @@
+  contentType: Option[String] = None
+)
+
+object UploadOptions {
```
you can remove this companion object, dont need it

melgenek replied:
done in this commit https://github.com/zio/zio-s3/pull/117/commits/bda1d842c0e9c8f75efb4e3e1084fc3270b79a96#diff-362a387518f51a267b646add5682a0d6R33

regis-leray commented on src/main/scala/zio/s3/MultipartUploadOptions.scala:
```diff
@@ -0,0 +1,35 @@
+)
+
+object MultipartUploadOptions {
+  final val MinMultipartPartSize: Int = 5 * 1024 * 1024 // 5 MB
```
could you please rename MinChunkSize

melgenek replied:
https://github.com/zio/zio-s3/pull/117/commits/bda1d842c0e9c8f75efb4e3e1084fc3270b79a96#diff-362a387518f51a267b646add5682a0d6R33

regis-leray commented on src/main/scala/zio/s3/MultipartUploadOptions.scala:
```diff
@@ -0,0 +1,35 @@
+object MultipartUploadOptions {
+  final val MinMultipartPartSize: Int = 5 * 1024 * 1024 // 5 MB
+
+  final val default = MultipartUploadOptions()
```
you can remove it since we are using default values

melgenek replied:
done in this commit https://github.com/zio/zio-s3/pull/117/commits/bda1d842c0e9c8f75efb4e3e1084fc3270b79a96

regis-leray commented on src/main/scala/zio/s3/Live.scala:
```diff
@@ -105,47 +108,58 @@ final class Live(unsafeClient: S3AsyncClient) extends S3.Service {
-                    _.createMultipartUpload(
-                      CreateMultipartUploadRequest
+                    _.createMultipartUpload {
+                      val builder = CreateMultipartUploadRequest
```
you can remove this `val builder`

melgenek replied:
how? The content type is optional, that's why there is a builder that is then optionally updated with a content type.

regis-leray replied:
oups my bad all good 👍

regis-leray commented:
@melgenek thank you so much, great contribution ! you improved the performance :)
one little commentt and we are good to merge

melgenek commented:
Cool :smile:! Thank you, @regis-leray!

melgenek commented:
@regis-leray is there a chance that you could release this over the weekend? I would love to use this update in my codebase.
It seems that this change deserves a minor version bump :smile:

regis-leray commented:
Of course i will. I want to do some cleanup. 
You can use the snapshot for now.
